server/netty-server/src/main/scala/sttp/tapir/server/netty/internal/ReactiveWebSocketHandler.scala

```diff
@@ -0,0 +1,182 @@
+    isSsl: Boolean
+)(implicit m: MonadError[F])
+    extends ChannelInboundHandlerAdapter {
+
+  // By using the Netty event loop assigned to this channel we get two benefits:
+  //  1. We can avoid the necessary hopping around of threads since Netty pipelines will
+  //     only pass events up and down from within the event loop to which it is assigned.
+  //     That means calls to ctx.read(), and ctx.write(..), would have to be trampolined otherwise.
+  //  2. We get serialization of execution: the EventLoop is a serial execution queue so
+  //     we can rest easy knowing that no two events will be executed in parallel.
+  private[this] var eventLoopContext: ExecutionContext = _
+
+  private val logger = LoggerFactory.getLogger(getClass.getName)
+  private val WebSocketAutoPingHandlerName = "wsAutoPingHandler"
+
+  def isWsHandshake(req: HttpRequest): Boolean =
+    "Websocket".equalsIgnoreCase(req.headers().get(HttpHeaderNames.UPGRADE)) &&
+      "Upgrade".equalsIgnoreCase(req.headers().get(HttpHeaderNames.CONNECTION))
+
+  override def handlerAdded(ctx: ChannelHandlerContext): Unit =
+    if (ctx.channel.isActive) {
+      initHandler(ctx)
+    }
+  override def channelActive(ctx: ChannelHandlerContext): Unit = {
+    channelGroup.add(ctx.channel)
+    initHandler(ctx)
+  }
+
+  private[this] def initHandler(ctx: ChannelHandlerContext): Unit = {
+    if (eventLoopContext == null)
+      eventLoopContext = ExecutionContext.fromExecutor(ctx.channel.eventLoop)
+  }
+
+  override def exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable): Unit = {
+    logger.error("Error while processing the request", cause)
+  }
+
+  override def channelRead(ctx: ChannelHandlerContext, msg: Any): Unit = {
+    def replyWithError500(reason: Throwable): Unit = {
+      logger.error("Error while processing the request", reason)
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.INTERNAL_SERVER_ERROR)
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, 0)
+      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+      val _ = ctx.writeAndFlush(res)
+    }
+
+    def rejectHandshakeForRegularEndpoint(content: NettyResponseContent): Unit = {
+      val message = "Unexpected WebSocket handhake on a regular HTTP endpoint"
+      val res = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1, HttpResponseStatus.BAD_REQUEST, Unpooled.wrappedBuffer(message.getBytes))
+      res.headers().set(HttpHeaderNames.CONTENT_LENGTH, message.length())
+      res.headers.set(HttpHeaderNames.CONNECTION, HttpHeaderValues.CLOSE)
+      content.channelPromise.setFailure(new IllegalStateException("Unexpected response content"))
+      val _ = ctx.writeAndFlush(res)
+    }
+
+    msg match {
+      case req: FullHttpRequest if isWsHandshake(req) =>
+        ctx.pipeline().remove(this)
+        ctx.pipeline().remove(classOf[ReadTimeoutHandler])
```

Comment:
so we don't need the constants after all? :)

Reply:
We need them for `pipeline.addAfter`, where you have to use names.